tests/unit/h_vialib/secure/token_test.py

```diff
@@ -0,0 +1,156 @@
+        (
+            InvalidSignatureError,
+            ExpiredSignatureError,
+            DecodeError,
+        ),
+    )
+    def test_verify_translates_errors(self, token, jwt, exception):
+        jwt.decode.side_effect = exception
+        with pytest.raises(InvalidToken):
+            token.verify("fake_token")
+
+    @pytest.fixture
+    def token(self):
+        return SecureToken("a_very_secret_secret")
+
+    @pytest.fixture
+    def jwt(self, patch):
+        return patch("h_vialib.secure.token.jwt")
+
+
+class TestSecureURLToken:
+    def test_create_normalizes_and_packs_the_url(self, token):
+        expires = datetime.now() + timedelta(seconds=10)
+
+        token_string = token.create("http://example.com", {"a": 2}, expires)
+
+        assert token_string == Any.string()
+        assert decode_token(token_string) == {
+            "url": "http://example.com/",
+            "a": 2,
+            "exp": Any.int(),
+        }
+
+    @pytest.mark.parametrize("url", (None, ""))
+    def test_create_requires_the_url(self, token, url):
+        with pytest.raises(ValueError):
+            token.create(url, {}, max_age=10)
+
+    def test_verify_works_with_a_matching_url(self, token):
+        token_string = token.create("http://example.com", {}, max_age=10)
+
+        decoded = token.verify(token_string, "http://example.com/")
+
+        assert decoded == Any.dict.containing({"url": "http://example.com/"})
+
+    def test_verify_fails_if_there_is_no_url_in_the_token(self, token):
+        no_url_token = SecureToken(token._secret).create({}, max_age=10)
+
+        with pytest.raises(InvalidToken):
+            token.verify(no_url_token, "http://any.example.com/")
+
+    def test_verify_fails_if_there_is_url_mismatch(self, token):
+        token_string = token.create("http://example.com", {}, max_age=10)
+
+        with pytest.raises(InvalidToken):
+            token.verify(token_string, "http://DIFFERENT.example.com/")
+
+    @pytest.mark.parametrize(
+        "url,normalized",
+        (
+            ("http://example.com/", "http://example.com/"),
+            ("http://example.com", "http://example.com/"),
+            ("http://example.com]", "http://example.com]"),
+        ),
+    )
+    @pytest.mark.parametrize("query_string", ("", "?a=b"))
+    def test_normalize_url(self, token, url, normalized, query_string):
+        result = token.normalize_url(url + query_string)
+
+        assert result == normalized + query_string
+
+    @pytest.fixture
+    def token(self):
+        return SecureURLToken("a_very_secret_secret")
+
+
+class TestViaSecureURLToken:
+    def test_create_works(self, token):
+        token_string = token.create("http://example.com?via.config=blah")
+
+        assert token_string == Any.string()
+        assert decode_token(token_string) == {
+            "url": "http://example.com/",
+            "exp": Any.int(),
+        }
+
+    def test_create_uses_a_quantized_expiry(self, token, quantized_expiry):
```

Comment:
This tests fails for me locally, shouldn't `quantized_expiry.return_value` be a bit in the future (vs just `now()`)?

Reply:
What failure are you getting?

Reply:
Sorry, looks like:
```AssertionError: assert {'exp': 1612787391, 'url': '*any*'} == <AnyDict '* any dict containing {'exp': 1612783791} *'>
E        +  where {'exp': 1612787391, 'url': '*any*'} = decode_token('eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1cmwiOiIqYW55KiIsImV4cCI6MTYxMjc4NzM5MX0.8nzPoBo9rz8xf0_mPgwGXqnpseWWvY889hpPNttmsFg')
E        +  and   <AnyDict '* any dict containing {'exp': 1612783791} *'> = <function fluent_entrypoint.__get__.<locals>.wrapper at 0x7f8bcf087d08>({'exp': 1612783791})
E        +    where <function fluent_entrypoint.__get__.<locals>.wrapper at 0x7f8bcf087d08> = <class 'h_matchers.matcher.collection.AnyDict'>.containing
E        +      where <class 'h_matchers.matcher.collection.AnyDict'> = Any.dict
```

Reply:
I don't think the future / past makes a difference here, as we aren't checking that the token is valid, just that it has the contents we expect. It might be nice from a demonstrating the behavior point of view.

I'll scrub out my local tox envs and give this another shot. The core guts of the things being compared look identical, but I'm not sure about all the trappings.

We should go through `h-matchers` at some point, as it's our own home grown special testing sauce. It's very handy some times, but hard to understand and occasionally makes debugging tests a pain.

Reply:
Also missing the elephant in the room, it's passing in CI too